### ISSUE TITLE
RFC: Bind directly to SFML with bindgen, bypassing CSFML

### ISSUE THREAD

**Issue by crumblingstatue** [jeremyletang/rust-sfml#194]
Update: This is on hold indefinitely, as I have ran into problems trying to implement it.

[CSFML](https://github.com/SFML/CSFML) is kind of slow moving. It always lags behind SFML releases. It still doesn't have an official 2.5 release.

We could bypass CSFML by using [bindgen](https://github.com/rust-lang-nursery/rust-bindgen), and update to the latest SFML at our own pace.

This would also bring possible performance improvements, as CSFML makes most SFML types opaque, and hides them behind a heap allocation. With bindgen, we could allocate these types directly on the stack. Update: Some of these types are self-referential, and Rust lacks support for these kinds of types, so we would likely need to put them behind heap allocations anyway.

However, there are some disadvantages to this method:

- Bindgen uses libclang to get the necessary information to bind C++. This means to build rust-sfml, the user has to have libclang installed.
- Most likely we would still need to create a thin C++ -> C wrapper for a few API elements. These possibly include functions that take `std::string` as arguments. We would need to create a CSFML-like wrapper that takes `const char *`, but internally calls the `std::string` constructor. This means that to build rust-sfml, the user also has to have a C++ compiler installed.
- The above also mean that building rust-sfml will be (possibly a lot) slower, and fast compilation was one of the main advantages of rust-sfml over pure rust libraries.

I personally would like to eventually go through with this, because I don't like waiting on CSFML, and I want rust-sfml to be as efficient as possible, and the extra allocations bug me.

But I would like to gather public opinion on this, as this is a heavy change that could make rust-sfml unappealing to some users.

In the meantime, I will do some experimenting over at https://github.com/crumblingstatue/rust-sfml-direct.

**Comment by crumblingstatue** [comment]
I'm having serious troubles binding `sf::Music`. The music playing thread is hanging for some reason and there is other odd behavior. I don't know if this is a bindgen bug (maybe the `sf::Thread` and `sf::Mutex` inside the `sf::SoundStream` inside `sf::Music` have the wrong layout?), or if I'm doing something wrong, but even if no one objects to my above plans, this probably won't come to fruition for a long time.

**Comment by crumblingstatue** [comment]
Okay, turns out `sf::Music` is self-referential, and bindgen's auto-generated `new()` function moves it, and that causes the weird behavior. I'll have to think about how to solve this.

**Comment by crumblingstatue** [comment]
I'll put this on hold indefinitely, until the support for self-referential types is improved in Rust.
